File: 백준/Gold/2589. 보물섬/보물섬.py
from collections import deque
import sys
input = sys.stdin.readline

N, M = map(int, input().split())
maps = []
for _ in range(N):
    maps.append(list(input().strip()))

def createVisited():
    return [[0] * M for _ in range(N)]

def goingNoGoing(nx, ny, visited):

    # maps 안에 좌표가 있고, 육지이고, 방문하지 않았으면 True 리턴
    if 0 <= nx < N and 0 <= ny < M and maps[nx][ny] == 'L' and visited[nx][ny] == 0:
        return True
    return False

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]

def bfs(i, j):

    queue = deque()
    queue.append((i, j))

    visited = createVisited()
    visited[i][j] = 1

    count = 0

    while queue:
        x, y = queue.popleft()

        # 상하좌우 4방향에 대하여 탐색
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            
            # 갈 수 있는 곳일 경우
            if goingNoGoing(nx, ny, visited):
                visited[nx][ny] = visited[x][y] + 1
                count = max(count, visited[nx][ny])  # 최대 거리 갱신
                queue.append((nx, ny))  # 큐에 넣기

    # 시작 지점 값 1부터 시작했으므로, 1을 뺀 값을 return 한다
    return count - 1

result = 0

for i in range(N):
    for j in range(M):
        # 모든 육지 좌표에 대하여 bfs 진행
        if maps[i][j] == 'L':
            result = max(result, bfs(i, j))

print(result)

File: 백준/Gold/2589. 보물섬/보물섬.java
import java.io.*;
import java.util.*;

public class Main {
    static int N, M;
    static char[][] maps;
    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, -1, 1};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        
        maps = new char[N][M];
        for (int i = 0; i < N; i++) {
            maps[i] = br.readLine().toCharArray();
        }
        
        int result = 0;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (maps[i][j] == 'L') {
                    result = Math.max(result, bfs(i, j));
                }
            }
        }
        
        System.out.println(result);
    }

    static boolean goingNoGoing(int nx, int ny, int[][] visited) {
        return nx >= 0 && nx < N && ny >= 0 && ny < M && maps[nx][ny] == 'L' && visited[nx][ny] == 0;
    }

    static int bfs(int i, int j) {
        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[]{i, j});
        
        int[][] visited = new int[N][M];
        visited[i][j] = 1;
        
        int count = 0;
        
        while (!queue.isEmpty()) {
            int[] current = queue.poll();
            int x = current[0];
            int y = current[1];
            
            for (int k = 0; k < 4; k++) {
                int nx = x + dx[k];
                int ny = y + dy[k];
                
                if (goingNoGoing(nx, ny, visited)) {
                    visited[nx][ny] = visited[x][y] + 1;
                    count = Math.max(count, visited[nx][ny]);
                    queue.add(new int[]{nx, ny});
                }
            }
        }
        
        return count - 1;
    }
}